src/equations/ideal_glm_mhd_multiion.jl

```diff
@@ -263,4 +279,165 @@ end
 
     return SVector(cons)
 end
+
+@doc raw"""
+    source_terms_collision_ion_ion(u, x, t,
+                                   equations::AbstractIdealGlmMhdMultiIonEquations)
+
+Compute the ion-ion collision source terms for the momentum and energy equations of each ion species as
+```math
+\begin{aligned}
+  \vec{s}_{\rho_k \vec{v}_k} =&  \rho_k\sum_{k'}\bar{\nu}_{kk'}(\vec{v}_{k'} - \vec{v}_k),\\
+  s_{E_k}  =& 
+    3 \sum_{k'} \left(
+    \bar{\nu}_{kk'} \frac{\rho_k M_{1}}{M_{k'} + M_k} R_1 (T_{k'} - T_k)
+    \right) + 
+    \sum_{k'} \left(
+        \bar{\nu}_{kk'} \rho_k \frac{M_{k'}}{M_{k'} + M_k} \norm{\vec{v}_{k'} - \vec{v}_k}^2
+        \right)
+        +
+        \vec{v}_k \cdot \vec{s}_{\rho_k \vec{v}_k},
+\end{aligned}
+```
+where ``M_k`` is the molar mass of ion species `k` provided in `molar_masses`, 
+``R_k`` is specific gas constant of ion species `k` provided in `gas_constants`, and
+ ``\bar{\nu}_{kk'}`` is the effective collision frequency of species `k` with species `k'`, which is computed as
+```math
+\begin{aligned}
+  \bar{\nu}_{kk'} = \bar{\nu}^1_{kk'} \tilde{B}_{kk'} \frac{\rho_{k'}}{T_{k k'}^{3/2}},
+\end{aligned}
+```
+with the so-called reduced temperature ``T_{k k'}`` and the ion-ion collision constants ``\tilde{B}_{kk'}`` provided
+in `ion_electron_collision_constants`.
+
+The additional coefficient ``\bar{\nu}^1_{kk'}`` is a non-dimensional drift correction factor proposed by Rambo and 
+Denavit.
+
+References:
+- P. Rambo, J. Denavit, Interpenetration and ion separation in colliding plasmas, Physics of Plasmas 1 (1994) 4050–4060.
+- Schunk, R. W., Nagy, A. F. (2000). Ionospheres: Physics, plasma physics, and chemistry. 
+  Cambridge university press.
+"""
+function source_terms_collision_ion_ion(u, x, t,
+                                        equations::AbstractIdealGlmMhdMultiIonEquations)
+    s = zero(MVector{nvariables(equations), eltype(u)})
+    @unpack gas_constants, molar_masses, ion_ion_collision_constants = equations
+
+    prim = cons2prim(u, equations)
+
+    for k in eachcomponent(equations)
+        rho_k, v1_k, v2_k, v3_k, p_k = get_component(k, prim, equations)
+        T_k = p_k / (rho_k * gas_constants[k])
+
+        S_q1 = 0
+        S_q2 = 0
+        S_q3 = 0
+        S_E = 0
+        for l in eachcomponent(equations)
+            rho_l, v1_l, v2_l, v3_l, p_l = get_component(l, prim, equations)
```

Comment:
Is there a good way to skip the case `k == l` without causing allocations?
I have tried the following versions of the code, but they all cause allocations:
* Version 1:
```julia
for l in eachcomponent(equations)
    l == k && continue
    ....
end
```
* Version 2:
```julia
for l in eachcomponent(equations)
    if l != k
        ...
    end
end
```
* Version 3:
```julia
iter = Iterators.flatten((1:k - 1, k + 1:ncomponents(equations)))
for l in iter
    ...
end
```
* Version 4:
```julia
iter = Iterators.filter(i -> i != k, 1:ncomponents(equations))
for l in iter
    ...
end
```

Reply:
What happens if you do not initialize the `S_q{123}` an `S_E` values to `zero(eltype(u))` instead of `0`? This is the only obvious type instability I can see occurring here, since all other variables seem to be local to the inner loop.